shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -20,60 +20,104 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+    bodyTy: SimpleType,
     mthDecls: List[MethodDef[Right[Term, Type]]],
     mthDefs: List[MethodDef[Left[Term, Type]]],
     baseClasses: Set[Var],
     toLoc: Opt[Loc],
   ) {
     def allBaseClasses(ctx: Ctx)(implicit traversed: Set[Var]): Set[Var] =
-      baseClasses.map(v => Var(v.name.toList.mapHead(_.toLower).mkString)) ++
+      baseClasses.map(v => Var(v.name.decapitalize)) ++
         baseClasses.iterator.filterNot(traversed).flatMap(v =>
           ctx.tyDefs.get(v.name).fold(Set.empty[Var])(_.allBaseClasses(ctx)(traversed + v)))
-    val targsMap: Map[Str, TypeVariable] = tparams.map(_.name).zip(targs).toMap
+    val (tparams: List[TypeName], targs: List[TypeVariable]) = tparamsargs.unzip
+    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
+    def wrapMethod(pt: PolymorphicType)(implicit prov: TypeProvenance): MethodType =
+      MethodType(pt.level, S(FunctionType(singleTup(thisTy), pt.body)(prov)), nme)
+    def wrapMethod(mt: MethodType)(implicit prov: TypeProvenance): MethodType =
+      if (mt.body.nonEmpty) wrapMethod(mt.toPT) else mt.copy(parents = nme :: Nil)
+  }
+
+  private case class MethodDefs(
+    tn: TypeName,
+    parents: List[MethodDefs],
+    decls: Map[Str, MethodType],
+    defns: Map[Str, MethodType],
+  ) {
+    private def &(that: MethodDefs)(tn: TypeName, newdefns: Set[Str])(implicit raise: Raise): MethodDefs =
+      MethodDefs(
+        tn,
+        this.parents ::: that.parents,
+        mergeMap(this.decls, that.decls)(_ & _),
+        (this.defns.iterator ++ that.defns.iterator).toSeq.groupMap(_._1)(_._2).flatMap {
+          case _ -> Nil => die
+          case mn -> (defn :: Nil) => S(mn -> defn)
+          case mn -> defns if newdefns(mn) => N
+          case mn -> defns =>
+            implicit val prov: TypeProvenance = TypeProvenance(tn.toLoc, "inherited method declaration")
+            err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
+              :: msg"Definitions of method $mn inherited from:" -> N
+              :: defns.iterator.map(mt => msg"• ${mt.parents.head}" -> mt.prov.loco).toList)
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil))
+        }
+      )
+    private def ++(that: MethodDefs): MethodDefs = {
+      require(parents.isEmpty)
+      MethodDefs(
+        that.tn,
+        that.parents,
+        this.decls ++ that.decls,
+        this.defns ++ that.defns
+      )
+    }
+    private def addParent(prt: TypeName): MethodDefs =
+      copy(decls = decls.view.mapValues(mt => mt.copy(parents = prt :: mt.parents)).toMap)
+    def propagate(top: Bool = true)(implicit raise: Raise): MethodDefs =
+      parents.map(_.propagate(false)).reduceOption(_.&(_)(tn, defns.keySet)).map(_.addParent(tn))
+        .foldRight(if (top) copy(parents = Nil, decls = Map.empty, defns = Map.empty) else copy(parents = Nil))(_ ++ _)
```

Comment:
This `copy` call may be cleaner as:
```suggestion
        .foldRight(if (top) MethodDefs(tn, Nil, Map.empty, Map.empty) else copy(parents = Nil))(_ ++ _)
```

Reply:
Addressed in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8. (I wrote it this way to make the two branches more symmetric.)